managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2532,10 +2532,45 @@ void internalFlushPendingMarkDeletes() {
 
     void createNewMetadataLedger(final VoidCallback callback) {
         ledger.mbean.startCursorLedgerCreateOp();
+        doCreateNewMetadataLedger().thenAccept(newLedgerHandle -> {
+            if (newLedgerHandle == null) {
+                return;
+            }
+            MarkDeleteEntry mdEntry = lastMarkDeleteEntry;
+            // Created the ledger, now write the last position content
+            persistPositionToLedger(newLedgerHandle, mdEntry, new VoidCallback() {
+                @Override
+                public void operationComplete() {
+                    if (log.isDebugEnabled()) {
+                        log.debug("[{}] Persisted position {} for cursor {}", ledger.getName(),
+                                mdEntry.newPosition, name);
+                    }
+                    switchToNewLedger(newLedgerHandle, callback);
+                }
+
+                @Override
+                public void operationFailed(ManagedLedgerException exception) {
+                    log.warn("[{}] Failed to persist position {} for cursor {}", ledger.getName(),
+                            mdEntry.newPosition, name);
+
+                    deleteLedgerAsync(newLedgerHandle);
+                    callback.operationFailed(exception);
+                }
+            });
+        }).whenComplete((result, e) -> {
+            ledger.mbean.endCursorLedgerCreateOp();
+            if (e != null) {
+                callback.operationFailed(createManagedLedgerException(e));
+            }
+        });
+    }
 
+    private CompletableFuture<LedgerHandle> doCreateNewMetadataLedger() {
+        CompletableFuture<LedgerHandle> future = new CompletableFuture<>();
         ledger.asyncCreateLedger(bookkeeper, config, digestType, (rc, lh, ctx) -> {
 
             if (ledger.checkAndCompleteLedgerOpTask(rc, lh, ctx)) {
+                future.complete(null);
```

Comment:
Question:
Can we return an exception future there? Because returning null directly is a bit weird to other users and prone to throw NPE.

Reply:
The previous logic was that if the judgment here is true, this directly returns. I need to know whether the value is true. Therefore, null is returned. If the value is null, the metadata does not need to be persisted.
Or, do you have any advice

Reply:
How about moving this logic out of this method? This will make this method purer. and you don't need to judge conditions twice.

```java
if (ledger.checkAndCompleteLedgerOpTask(rc, lh, ctx)) {
      future.complete(null);
      return;
}
```

Reply:
How about moving this logic out of this method? This will make this method purer. and you don't need to judge conditions twice.
```java
if (ledger.checkAndCompleteLedgerOpTask(rc, lh, ctx)) {
        future.complete(null);
	return;
}
```

Reply:
If we move it out, it changes the previous behavior and executes the following logic even if the handle is not available:
`ledger.getExecutor().execute`

Reply:
Sorry, I missed the folding code for GH. Although I think it should be possible to optimize, there are too many race conditions here. It's best not to change the logic.

Very thanks for your patient answer.